spmd/api.py

```diff
@@ -0,0 +1,318 @@
+                                        dtn, lambda n: value_remap[n]
+                                    )
+
+                if has_partial:
+                    gm.graph.erase_node(node)
+                    gm.graph.output(new_args)
+                break
+        else:
+            raise ValueError(f"Unrecognized node {node}")
+
+    # replace nodes in local traced graph with DTensor's dispatch graph
+    for node in gm.graph.nodes:
+        if node not in replacements:
+            continue
+
+        traced_dispatch = replacements[node]
+        # Map DT's dispatch graph input placeholder nodes to the ones in
+        # local traced graph. It uses index-based accessing, which is
+        # brittle, just for testing purpose.
+        flatten_args, _ = tree_flatten(node.args)
+        i, value_remap = 0, {}
+        for dtn in traced_dispatch.graph.nodes:
+            if dtn.op == "placeholder":
+                value_remap[dtn] = flatten_args[i]
+                i += 1
+
+        # insert DT's dispatch graph to traced local graph.
+        with gm.graph.inserting_before(node):
+            for dtn in traced_dispatch.graph.nodes:
+                if dtn.op == "placeholder":
+                    # do nothing, ignore placeholders, as it has already
+                    # been prepared in value_remap
+                    pass
+                elif dtn.op == "output":
+                    assert (
+                        len(dtn.args) == 1 and len(dtn.args[0]) == 1
+                    ), f"Expecting single output, but got {dtn.args}"
+                    node.replace_all_uses_with(value_remap[dtn.args[0][0]])
+                else:
+                    value_remap[dtn] = gm.graph.node_copy(
+                        dtn, lambda n: value_remap[n]
+                    )
+
+    gm.graph.lint()
+    gm.graph.eliminate_dead_code()
+    gm.recompile()
+
+    return gm
+
+
+class SPMD(nn.Module):
+    # TODO: add schema_override
+    def __init__(self, module: nn.Module, schema: Schema):
+        super().__init__()
+        assert schema.placements == [
+            Replicate()
+        ], "SPMD only support Replicate() parameters for now"
+
+        # TODO: coalesce broadcasts
+        for p in module.parameters():
+            dist.broadcast(p, src=0)
+
+        self._local_module = module
+        self._schema = schema
+        self._compiled_m = None
+
+    def _compile(
+        self, gm: fx.GraphModule, inps: List[torch.Tensor]
+    ) -> fx.GraphModule:
+        def is_param(t: torch.Tensor) -> bool:
+            # N.B.: id(t) and id(param) does not match
+            return t.storage().data_ptr() in [
+                p.storage().data_ptr() for p in self._local_module.parameters()
+            ]
+
+        shard_schema = Schema(mesh=self._schema.mesh, placements=[Shard(0)])
+        schemas = [
+            self._schema if is_param(inp) else shard_schema for inp in inps
```

Comment:
This is ensuring that we shard input on the batch dimension for now right?

Reply:
yep